text/0309-weak-cross-stack-references.md

```diff
@@ -0,0 +1,194 @@
+
+To turn a strong reference into a weak reference, call `Reference.weakStringReference(...)` (or `weakListReference`, or `weakNumberReference`). For example, the following references a bucket name weakly:
+
+```ts
+const bucket = new s3.Bucket(this, 'MyBucket');
+
+// ...
+
+const fn = new lambda.Function(this, 'MyFunction', {
+  // ...
+
+  environment: {
+    BUCKET_NAME: Reference.weakStringReference(bucket.bucketName),
+  },
+});
+```
+
+`Reference.weakStringReference()` only has an effect if the reference is to a construct defined in a different stack: it will use SSM Parameter Store to transport the value between stacks. If the reference is to a construct in the same stack, nothing happens (do be aware that in the example above, if `Bucket` gets renamed, there is a certain time window in which the Lambda Function may continue to be executed with the old bucket name!).
+
+If the value passed to `Reference.weakStringReference()` contains multiple references (either because it's a string with multiple references or it's an array or complex data structure), all of them are made weak.
+
+### Switching from strong to weak references
+
+If you have previously deployed stacks with strong cross-stack references and want to switch them to weak references, you must do this in a two-phase deployment.
+
+First, pass `strongToWeak: true` for the reference you are converting from strong to weak. This will generate both the strong and weak exports in the producing stack, while only consuming the weak reference in the consuming stack:
+
+```ts
+// First deployment, generate both exports
+Reference.weakStringReference(bucket.bucketName, {
+  strongToWeak: true,
+})
+```
+
+After you have deployed your stack with `strongToWeak` enabled, you can remove the flag again and do another deployment:
+
+```ts
+// Second deployment, stop generating strong export
+Reference.weakStringReference(bucket.bucketName)
+```
+
+### Defaults for stacks
+
+You can choose to configure the default reference style for an entire stack passing the `defaultReferenceStyle` property to a stack constructor:
+
+```ts
+new MyStack(this, 'MyStack', {
+  defaultReferenceStyle: ReferenceStyle.STRONG,
+  // or
+  // defaultReferenceStyle: ReferenceStyle.WEAK,
+  // defaultReferenceStyle: ReferenceStyle.STRONG_TO_WEAK,
+});
+```
+
+It is recommended to select reference style on a reference-by-reference basis, and not use this method, but you can if you want to.
+
+### Defaults for dashboards
+
+Strong references cause a lot of usability problems for CloudWatch Dashboards. That's why starting in new applications, all CloudWatch Dashboards will use weak references by default.
+
+This is controlled by the feature flag `@aws-cdk/aws-cloudwatch:dashboardWeakReferences`. It can take on the values `true` (for weak references) or `"strongToWeak"`, for the first phase of the two-phase deployment.
```

Comment:
What case can be made that **not** switching to weak references in CloudWatch Dashboards is the right thing to do?

Reply:
I don't think there is any. Strong references only make sense if the consumer would suffer availability impact if the producer disappears. This is definitely not the case for dashboards: the line will just have a gap in it.

Note: explicitly *not* saying this for *Alarms*: they may have impact if the producer disappears, depending on whether the `MissingDataPoints` property is set to `BREACHING` or `IGNORE`.